### PR TITLE
[Android] Implement cookies

### PR DESCRIPTION
**Pull request by nucleartux**

### CODE REVIEW

**Comment by mkonicek** [comment]
How have you tested this? Can you extend an example in the UI Explorer showing how this works?

**Comment by nucleartux** [comment]
Do you have ideas how I can show how it works? In my project I tested this on authentication on internal site using fetch API.

**Comment by mkonicek** [comment]
Ideally you shuold come up with an exhaustive set of examples showing this implements the Same origin policy: https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy

There should be an example in the UI Explorer demonstrating this, here are instructions on how to run it:
https://github.com/facebook/react-native/tree/master/Examples/UIExplorer

**Comment by ide** [comment]
Does SOP make sense here? What is the value of the origin?

**Comment by mkonicek** [comment]
I believe this is to fix https://github.com/facebook/react-native/issues/2792

Based the discussion there it looks like we should honor `credentials: 'same-origin'` in the `fetch` / XHR API like we do on iOS.

**Comment by nucleartux** [comment]
@mkonicek I added example, please review

**Comment by mkonicek** [comment]
@ide I'm not sure I understand - in the browser, XHR respects the same origin policy, correct?
From https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy:

    The same-origin policy controls interactions between two different origins, such as when you use XMLHttpRequest

Thanks for the example. I think we really need to demonstrate here that the implementation respects the same origin policy. I'm not sure `CookiePolicy.ACCEPT_ALL` does that. What does `ACCEPT_ORIGINAL_SERVER` do? http://docs.oracle.com/javase/7/docs/api/java/net/CookiePolicy.html

I think you'd need to send a request and get back cookies, then send **another** request to the same domain and demonstrate those cookies were sent as part of the second request.

Then you'd send a third request to a different domain (and ideally to a multiple of them) and demonstrate the cookies were not sent.

This should really be an integration test and we'll work on open sourcing those so you can add them, but an example would be a good start.

Note I'll be away for a week and will continue reviewing when I'm back.

**Comment by ide** [comment]
@mkonicek the same-origin policy compares the current page's origin with the origin of the request endpoint. If google.com does a fetch request to google.com, then `credentials: 'same-origin'` will include the google.com cookies. Also if google.com sends a fetch request to facebook.com, it will not send the facebook.com cookies unless `credentials: 'include'` is set. The other part of the SOP is that google.com cannot read the response from facebook.com unless special CORS headers are set.

Anyway, with RN the app doesn't have an origin, so I don't know what it means to enforce the SOP. There are just two settings: either send the cookies or don't. (maybe there's a third option of whether to send cookies if there's a 3xx redirect?)

**Comment by ide** [comment]
I think there may be some confusion -- when an XHR makes a request to facebook.com, it never should send the google.com cookies under any circumstance (this is separate from the SOP).

**Comment by nucleartux** [comment]
I readed RFCs and I think @mkonicek is right. Changed policy to more secure ACCEPT_ORIGINAL_SERVER.

**Comment by sooth-sayer** [comment]
+1

**Comment by sooth-sayer** [comment]
@ide @mkonicek Could you review this PR?

**Comment by gadiel** [comment]
+1

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 1]
```diff
@@ -56,6 +60,9 @@
     mClient = client;
     mShuttingDown = false;
     mDefaultUserAgent = defaultUserAgent;
+    mCookieManager = new CookieManager();
+    mCookieManager.setCookiePolicy(CookiePolicy.ACCEPT_ORIGINAL_SERVER);
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: This is the default (https://docs.oracle.com/javase/7/docs/api/java/net/CookieManager.html#setCookiePolicy(java.net.CookiePolicy)) but doesn't hurt to be explicit here

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 2]
```diff
@@ -47,6 +50,7 @@
   private final OkHttpClient mClient;
   private final @Nullable String mDefaultUserAgent;
   private boolean mShuttingDown;
+  private final CookieManager mCookieManager;
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: can you move this under the mClient declaration?

**Comment by ide** [comment]
From my research on OkHttpClient and cookie handling this looks like a good first pass. There are later features we could consider like a cookie store although that raises questions about being able to clear it, control separate instances of it (sometimes two NetworkingModule instances should share the store, sometimes you may not want them to), etc. But for basic session functionality, this looks good.

**Comment by mkonicek** [comment]
> when an XHR makes a request to facebook.com, it never should send the google.com cookies under any circumstance (this is separate from the SOP).

Yes this is what I meant by "SOP", sorry about the confusion.

@nucleartux Can you verify this is the case? You mentioned:

> Changed policy to more secure ACCEPT_ORIGINAL_SERVER.

But don't see this change.

**Comment by ide** [comment]
ACCEPT_ORIGINAL_SERVER sounds like it prevents Facebook.com from setting cookies for Google.com (also important!!).

I'm not sure if there is an explicit setting to tell the client to send only the Facebook.com cookies to Facebook.com. We should verify but the http client might always do that by default.

**Comment by nucleartux** [comment]
@mkonicek this is default policy

**Comment by lexs** [comment]
I've been working on this internally but I think we want to merge this just to have some support. As long as the policy is ACCEPT_ORIGINAL_SERVER `CookieManager` will behave as expected. The policy refers to that it accepts cookies where the domain in the cookie matches that of the request uri.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 3]
```diff
@@ -45,6 +48,7 @@
   private static final String USER_AGENT_HEADER_NAME = "user-agent";
 
   private final OkHttpClient mClient;
+  private final CookieManager mCookieManager;
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: You can skip this field and just construct in inline. `mClient.setCookieHandler(new CookieManager());`

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 4]
```diff
@@ -14,6 +14,9 @@
 import java.io.IOException;
 import java.io.InputStream;
 
+import java.net.CookieManager;
+import java.net.CookiePolicy;
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: This import is no longer used. Kill it.

**Comment by sooth-sayer** [comment]
Is there any chance that this PR would be in 0.15 release?

**Comment by ide** [comment]
@sooth-sayer it will likely go out with 0.16.

**Comment by mkonicek** [comment]
> this is default policy

IMHO it's always better to be explicit, you clearly state the intention and defaults can change.

**Comment by mkonicek** [comment]
@lexs Thanks for looking at this PR!

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 4 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: It's used now :) Better to specify the policy explicitly.

**Comment by CCoffie** [comment]
+1  - Any updates on the progress of this? I'm working on a project that requires the use of session to use the api. I would love to use react-native but without this feature it can't really be done.

**Comment by nicklockwood** [comment]
ping @mkonicek

**Comment by marcusroberts** [comment]
:+1:  another vote for this, as it's an instant show stopper when trying to run my iOS apps on Android as all are based on Rails API servers with cookie sessions.

**Comment by mkonicek** [comment]
Sorry for the delay - I have an update: @lexs started working on a different approach in the meantime - one that shares the cookies between okhttp and the Android WebView.

**Comment by lexs** [comment]
Closing this out as https://github.com/facebook/react-native/commit/274c5c78c4933a09335fca0e2fdc266c80fe4017 just landed. Thanks for you effort @nucleartux. Hopefully this fixes your issue otherwise feel free to open another issue.

**Comment by 6i6arka** [comment]
Will applying this fix mean that `fetch` will set cookies automatically?

**Comment by lexs** [comment]
@6i6arka For https://github.com/facebook/react-native/commit/274c5c78c4933a09335fca0e2fdc266c80fe4017 yes. With this PR yes, but only in-memory.

**Comment by wdyfantasy** [comment]
is cookies can be used in android?

**Pull request closed without merging**